### PR TITLE
LCOW: Service VM lifetime changes

### PR DESCRIPTION
**Pull request by lowenna**
Signed-off-by: John Howard <jhoward@microsoft.com>
    
This changes the LCOW driver to support both global SVM lifetime and per-instance lifetime, and makes the driver thread-safe. It corrects a number of previously existing bugs. It also bumps to the latest version of jhowardmsft/opengcs.

Have validated basic CI internally here (building/pulling/committing/running/...) in both modes and validated there are no regressions from the previous behaviour

@johnstep PTAL.

### CODE REVIEW

**Comment by jhowardmsft** [comment]
Fixed golint errors

**Comment by jhowardmsft** [comment]
More golint fixes to follow tomorrow....

**Comment by jhowardmsft** [comment]
`go vet` errors fixed. Should be green now in CI

**Comment by vieux** [comment]
ping @johnstep :D

**Comment by johnstep** [comment]
@vieux Sorry, I had been reviewing some separately with @jhowardmsft and we got some lock issues addressed. I will finish up the review soon; so far, so good.

**Review by johnstep — APPROVED** [review]
LGTM on green, after changes from external review; thanks @jhowardmsft for real-time review.

**Comment by jhowardmsft** [comment]
Is green :)

**Comment by jhowardmsft** [comment]
@thaJeztah Who else could review this? (Windows only code)

**Comment by thaJeztah** [comment]
I'll have a look if there's someone available; maybe @StefanScherer is also interested to have a look

**Comment by jhowardmsft** [comment]
Rebased. @darrenstahlmsft Could you take a look?

File: daemon/graphdriver/lcow/lcow.go
Comment (darrenstahlmsft) [inline daemon/graphdriver/lcow/lcow.go]: Debug log indicating mutex is being released, but unless I'm missing something it's still held.

File: daemon/graphdriver/lcow/lcow.go
Comment (darrenstahlmsft) [inline daemon/graphdriver/lcow/lcow.go]: nit: This could be a defer unlock above

**Review by darstahl — COMMENTED** [review]
Not a maintainer, but LGTM other than one missing Unlock

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Yes, good catch. Thanks. Updated.

**Review by darstahl — APPROVED** [review]

File: daemon/graphdriver/lcow/lcow.go
Comment (darrenstahlmsft) [inline daemon/graphdriver/lcow/lcow.go]: RE offline discussion, this is fine as is for now.

File: daemon/graphdriver/lcow/lcow.go
Comment (darrenstahlmsft) [inline daemon/graphdriver/lcow/lcow.go]: 👍

**Comment by jhowardmsft** [comment]
@thaJeztah @Vieux - OK with the above reviews with Windows-eyes on it? I need to build on this code base, hence would like it to get merged sooner, and similarly does Akash for the remote filesystem work. Thanks.

File: daemon/graphdriver/lcow/lcow.go
Comment (thaJeztah) [inline daemon/graphdriver/lcow/lcow.go]: actually, can we change this option to `--storage-opt lcow.globalmode=1`?

Storage-driver specific options use a `<driver>.` prefix (see https://github.com/docker/cli/blob/ff2552f7a1073834e79793aeb5dc5a4e486956a4/docs/reference/commandline/dockerd.md#options-per-storage-driver)

File: daemon/graphdriver/lcow/lcow.go
Comment (thaJeztah) [inline daemon/graphdriver/lcow/lcow.go]: Same here; actually, can you change this option to `lcow.globalmode`?

Storage-driver specific options use a `<driver>.` prefix (see https://github.com/docker/cli/blob/ff2552f7a1073834e79793aeb5dc5a4e486956a4/docs/reference/commandline/dockerd.md#options-per-storage-driver)

Comment (thaJeztah) [inline daemon/graphdriver/lcow/lcow.go]: Also can you make this a boolean instead of `1` / `0`, and check the value here? `--storage-opt lcow.globalmode=false` should disable it. This is important, because we use these inside the `daemon.json` as well (for which we use booleans)

**Review by thaJeztah — CHANGES_REQUESTED** [review]
Started reviewing, and left some comments

File: daemon/graphdriver/lcow/lcow.go
Comment (thaJeztah) [inline daemon/graphdriver/lcow/lcow.go]: Not necessarily for this PR ^^

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Before the likes of you start reading the comments too closely :gb:

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Sure, will move to above package.

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: See my comment 1 line higher....

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Sure will be in next update.

Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: BTW - documentation is going to come later. Let's get it working first.... docs will be a combined effort which @patricklang will have to take on the majority of

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Sure. Updated

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Yup, `lcow.globalmode`

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Yeah, let me do that in the current branch I'm working in for a future PR which builds on the lifetime changes in this PR. Will avoid a confusing rebase.

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: BTW - the statement about the prefix isn't universally true - see `btrfs.go` L569 for example.

Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: Updated to true/false using `strconv.ParseBool`

**Comment by jhowardmsft** [comment]
@thaJeztah Comments addressed, update pushed. Will address the return parameters in the follow-up PR once this is merged as mentioned ^^

**Comment by jhowardmsft** [comment]
@thajeztah and still green 💚

**Review by thaJeztah — APPROVED** [review]
discussing with @johnstep - we're good merging this; it's clearly WIP code, and this requires refactoring before it's reaching its final stage.


"lgtm"

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: That's fair. Yes, the debugging stuff is deliberately paranoid right now. Once it's stable, I can make changes such as this.

File: daemon/graphdriver/lcow/lcow.go
Comment (jhowardmsft) [inline daemon/graphdriver/lcow/lcow.go]: I'll update the comment. To something more like `there are no more references if we have reached this point`. Is that what you mean?

Comment (thaJeztah) [inline daemon/graphdriver/lcow/lcow.go]: Yes, or "don't" return after decrementing, but checking if `referenceCount` reaches `0`. May want to combine that with the refactoring; for now, updating the comment should do

**Pull request merged**
